File: basic_dsa_three_hundred_ninety_five.java
/**
 * Basic DSA Problems - Longest Substring with At Least K Repeating (LeetCode 395)
 * Longest substring where every char appears >= k times
 */
public class basic_dsa_three_hundred_ninety_five {

    // Problem: s, k. Longest substring with each char freq >= k.

    /**
     * Divide at chars with freq < k. Recurse on parts.
     * Time O(n), Space O(n)
     */
    public static int longestSubstring(String s, int k) {
        if (s.length() < k) return 0;
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        int bad = -1;
        for (int i = 0; i < 26; i++)
            if (freq[i] > 0 && freq[i] < k) { bad = i; break; }
        if (bad == -1) return s.length();
        int max = 0;
        for (String part : s.split(Character.toString((char) ('a' + bad))))
            max = Math.max(max, longestSubstring(part, k));
        return max;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Longest Substring with At Least K Repeating");
        System.out.println("=".repeat(60));

        System.out.println("aaabb, k=3 -> " + longestSubstring("aaabb", 3));
        System.out.println("ababbc, k=2 -> " + longestSubstring("ababbc", 2));
    }
}

File: basic_dsa_three_hundred_ninety_five.cpp
/**
 * Basic DSA - H-Index II (LeetCode 275)
 * Citations sorted ascending; compute h-index in O(log n).
 */
#include <iostream>
#include <string>
#include <vector>

int hIndex(std::vector<int>& citations) {
    int n = static_cast<int>(citations.size());
    int lo = 0;
    int hi = n;
    while (lo < hi) {
        int mid = (lo + hi + 1) / 2;
        if (citations[n - mid] >= mid)
            lo = mid;
        else
            hi = mid - 1;
    }
    return lo;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: H-Index II\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {0, 1, 3, 5, 6};
    std::cout << "[0,1,3,5,6] -> " << hIndex(v1) << "\n";

    std::vector<int> v2 = {1, 2, 100};
    std::cout << "[1,2,100] -> " << hIndex(v2) << "\n";
}
